modules/drivers/oracle/src/metabase/driver/oracle.clj

```diff
@@ -317,6 +319,53 @@
   [driver _ field-or-value]
   (sql.qp/unix-timestamp->honeysql driver :seconds (hx// field-or-value (hsql/raw 1000000))))
 
+(defn- time-zoned-trunc
+  "Same as [[trunc]], but truncates `x` to `unit` in the results timezone
+   if `x` is a timestamp with time zone."
+  [unit x]
+  (let [x (cond-> x
+             (hx/is-of-type? x #"(?i)timestamp(\(\d\))? with time zone")
+             (hx/->AtTimeZone (qp.timezone/results-timezone-id)))]
+    (trunc unit x)))
+
+(defmethod sql.qp/datetime-diff [:oracle :year]
+  [driver _unit x y]
+  (hx// (sql.qp/datetime-diff driver :month x y) 12))
+
+(defmethod sql.qp/datetime-diff [:oracle :quarter]
+  [driver _unit x y]
+  (hx// (sql.qp/datetime-diff driver :month x y) 3))
```

Comment:
I'm a little late to this party but I don't think quarters can be treated like this. A quarter isn't an arbitrary 3 month period. Imagine if there was a datetime-diff for "season". Dates in spring are one season away from summer. But you can't say that any date within three months of another date is in the same season.

Reply:
Fair point, it differs from how we often talk about quarters. But if we are to treat quarters consistently with years, months, weeks, then a quarter indeed should be an arbitrary period of three months. This is "whole unit" type calculation. I discussed this with Bruno when I first started implementing quarters.
https://metaboat.slack.com/archives/C03SHMZ266R/p1670413998743079?thread_ts=1670411552.316629&cid=C03SHMZ266R

Reply:
> it differs from how we often talk about quarters.

To recap, there are two ways this can work:

```clj
(floor (/ (- x y) 3))
```

or 

```clj
(- (quarter-of-year x) (quarter-of-year y))
```

I think other units like weeks or years clearly make more sense with the former behavior. e.g. I think `1` makes more sense as the answer the number of years between Dec 31st 2022 and Jan 1st 2024 than `2` does. For quarters, I think maybe it makes more sense to say there is `1` quarter between March 31st and July 1st rather than `2` quarters.

But this might be a product question. We should think about how people would actually use this functionality and which behavior would make more sense for our intended use cases.

If you have already considered this issue and discussed it with @brunobergher and settled on one behavior or another I don't want to second-guess that.

If we did want to switch the behaviors, it would actually be a simpler implementation I think:

```clj
(defmethod sql.qp/datetime-diff [:oracle :quarter]
  [driver unit x y]
  (- (sql.qp/date driver unit x) (sql.qp/date driver unit y)))
```

You could actually probably do that for all of the units.

CC @dpsutton

Reply:
I'm still not 100% sure this is the way to go, but we decided to err on the side of consistency with other unit calculations.

Reply:
@brunobergher Just to be super confident this is the right way to go, let's revisit this one more time. Because there's  context here that was left out of the [original conversation](https://metaboat.slack.com/archives/C01NT034X2B/p1666187460151969?thread_ts=1666026192.444749&cid=C01NT034X2B) on this.

A downside of counting whole period diffs instead of boundary diffs (what we're doing here) is that whole period diffs require much longer SQL expressions for year, quarter, and month for most DBs. MySQL and Postgres are the only DBs where their datediff-like function counts whole periods. 

For example, here's how a `datetimeDiff([column], [column], "year")` custom expression would look for bigquery:
```
(
 (
   datetime_diff(datetime(table.column), 
                 datetime(table.column), 
                 month)
   + 
   CASE 
     WHEN (datetime(table.column) < datetime(table.column)
       AND extract(day from table.column) > extract(day from table.column)) 
     THEN -1
     WHEN (datetime(table.column) > datetime(table.column)
       AND extract(day from table.column) < extract(day from table.column)) 
     THEN 1
     ELSE 0 
   END) 
 / 12
) AS years 
```

We need that huge case expression to adjust for the fact that bigquery's datetime_diff counts boundaries between months, not whole periods.

But if we changed our `datetimeDiff` to count boundaries, then the resulting SQL would be simply:

```
datetime_diff(datetime(table.column), datetime(table.column), month)
```

I'm not sure whether it's worth thinking about, because it really only affects users if they convert the question to SQL. But it's worth considering since it wasn't in the discussion before.

Reply:
I still strongly believe boundaries are not how humans think. We also model custom expressions off of Excel, and it counts complete entities.

Basically saying that these are the diffs between Dec 28 and Jan 3:
- Months: 1
- Years: 1
makes no practical sense.

Thanks for the double-check though.

Reply:
Great. Thanks